File: day1/704. 二分查找.cpp
// go:build ignore
#include <iostream>
using namespace std;
class Solution {
 public:
  int search(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size();  // 数组长度
    while (left < right) {
      int middle = left + ((right - left) >>
                           1);  // 左+半个区间，右移动1位=除以2,+-优先级高
      if (nums[middle] > target) {
        right = middle;  // right为右索引+1. 落在区间[left,middle)中
      } else if (nums[middle] < target) {
        left = middle + 1;  // 落在区间[middle + 1, right)
      } else {
        return middle;
      }
    }
    return -1;  // not existed
  }
};

class Solution {
 public:
  int search(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size() - 1;  // target in [left, right]

    while (left <= right) {
      int middle = left + ((right - left) >> 1);  // 左区间
      if (nums[middle] > target) {
        right = middle - 1;
      } else if (nums[middle] < target) {
        left = middle + 1;
      } else {
        return middle;
      }
    }
    return -1;  // not existed
  }
};

int main() {}

File: day1/704. 二分查找.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
    return nums.indexOf(target)
};
var search = function (nums, target) {
    // target in [left, right)
    let left = 0
    let right = nums.length
    while (left < right) {
        let middle = left + ((right - left) >> 1)
        if (nums[middle] > target) right = middle
        else if (nums[middle] < target) left = middle + 1
        else return middle
    }
    return -1;
};

var search = function (nums, target) {
    // target in [left, right]
    let left = 0
    let right = nums.length - 1
    while (left <= right) {
        middle = left + ((right - left) >> 1)
        if (nums[middle] > target) right = middle - 1
        else if (nums[middle] < target) left = middle + 1
        else return middle
    }
    return -1;
};